text/0099-otlp-http.md

```diff
@@ -0,0 +1,146 @@
+(OTLP/gRPC for short) and defines an additional way to use OTLP protocol over
+HTTP transport (OTLP/HTTP for short). OTLP/HTTP uses the same ProtoBuf payload
+that is used by OTLP/gRPC and defines how this payload is communicated over HTTP
+transport.
+
+OTLP/HTTP uses HTTP POST requests to send telemetry data from clients to
+servers. Implementations MAY use HTTP/1.1 or HTTP/2 transports. Implementations
+that use HTTP/2 transport SHOULD fallback to HTTP/1.1 transport if HTTP/2
+connection cannot be established.
+
+### Request
+
+Telemetry data is sent via HTTP POST request.
+
+The default URL path for requests that carry trace data is `/traces` (for
+example the full URL when connecting to "example.com" server will be
+`https://example.com/telemetry`). The request body is a ProtoBuf-encoded
+[`ExportTraceServiceRequest`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/trace/v1/trace_service.proto#L38)
+message.
+
+The default URL path for requests that carry metric data is `/metrics` and the
+request body is a ProtoBuf-encoded
+[`ExportMetricsServiceRequest`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/metrics/v1/metrics_service.proto#L35)
+message.
+
+The client MUST set "Content-Type: application/x-protobuf" request header. The
+client MAY gzip the content and in that case SHOULD include "Content-Encoding:
+gzip" request header. The client MAY include "Accept-Encoding: gzip" request
+header if it can receive gzip-encoded responses.
+
+Non-default URL paths for requests MAY be configured on the client and server
+sides.
+
+### Response
+
+#### Success
+
+On success the server MUST respond with `HTTP 200 OK`. Response body MUST be
+ProtoBuf-encoded
+[`ExportTraceServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/trace/v1/trace_service.proto#L47)
+message for traces and
+[`ExportMetricsServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/metrics/v1/metrics_service.proto#L44)
+message for metrics.
+
+The server MUST set "Content-Type: application/x-protobuf" response header. If
+the request header "Accept-Encoding: gzip" is present in the request the server
+MAY gzip-encode the response and set "Content-Encoding: gzip" response header.
+
+The server SHOULD respond with success no sooner than after successfully
+decoding and validating the request.
+
+#### Failures
+
+If the processing of the request fails because the request contains data that
+cannot be decoded or is otherwise invalid and such failure is permanent then the
+server MUST respond with `HTTP 400 Bad Request`. The client MUST NOT retry the
+request when it receives `HTTP 400 Bad Request` response.
```

Comment:
no payload explaining the issue?

Reply:
We don't have the equivalent in OTLP/gRPC either. You are right that it is a good idea to have one, especially if we know what we want to have in the payload (perhaps it is just an empty ProtoBuf message initially).
Let me think.

Reply:
in gRPC we at least had guidelines on return codes and using grpc error details for additional information. At minimum, it's customary to return some error message in the HTTP body, in addition to the status code.

Reply:
You are right, I forgot that it is possible to include error details in gRPC, not just the status code. Let me add here too.

Reply:
I wonder if there's already some standard around this, e.g. when using grpc-gateway to expose gRPC service over HTTP/JSON.

Reply:
grpc-gateway returns a JSON-encoded [Status](https://godoc.org/google.golang.org/genproto/googleapis/rpc/status#Status) message. Status contains `code` (which is superfluous, since it is already reflected in HTTP Status), `message` string and `details` (which is `Any` object).

We also already use the same `Status` message in OTLP/gRPC to report errors and use `details` to provide additional information (e.g. throttling interval).

While some of the info in `Status` ProtoBuf message is a duplicate of what is available in HTTP response (in HTTP status code and in HTTP headers) I am still inclined to use `Status` for consistency, instead of inventing yet another message format just for OTLP/HTTP error response bodies.

Thoughts?

Reply:
OK, adding.

Reply:
@yurishkuro Added, please have a look.